prometheus/src/lib.rs

```diff
@@ -6,40 +6,97 @@ mod snapshot_metrics;
+};
 use solana_sdk::pubkey::Pubkey;
-use std::{collections::HashSet, sync::Arc};
+use std::{
+    collections::HashSet,
+    sync::{Arc, RwLock},
+    thread,
+};
 
 #[derive(Clone, Copy)]
 pub struct Lamports(pub u64);
 
-pub fn render_prometheus(
-    banks_with_commitments: BanksWithCommitments,
-    cluster_info: &Arc<ClusterInfo>,
-    vote_accounts: &Arc<HashSet<Pubkey>>,
-    identity_config: &Arc<IdentityInfoMap>,
-    snapshot_config: &Option<SnapshotConfig>,
-) -> Vec<u8> {
-    // There are 3 levels of commitment for a bank:
-    // - finalized: most recent block *confirmed* by supermajority of the
-    // cluster.
-    // - confirmed: most recent block that has been *voted* on by supermajority
-    // of the cluster.
-    // - processed: most recent block.
-    let mut out: Vec<u8> = Vec::new();
-    bank_metrics::write_bank_metrics(&banks_with_commitments, &mut out).expect("IO error");
-    cluster_metrics::write_cluster_metrics(
-        &banks_with_commitments,
-        &cluster_info,
-        vote_accounts,
-        identity_config,
-        &mut out,
-    )
-    .expect("IO error");
-    if let Some(snapshot_config) = snapshot_config {
-        snapshot_metrics::write_snapshot_metrics(snapshot_config, &mut out).expect("IO error");
+pub struct PrometheusMetrics {
+    bank_forks: Arc<RwLock<BankForks>>,
+    block_commitment_cache: Arc<RwLock<BlockCommitmentCache>>,
+    cluster_info: Arc<ClusterInfo>,
+    vote_accounts: Arc<HashSet<Pubkey>>,
+    snapshot_config: Option<SnapshotConfig>,
+    /// Initialized based on vote_accounts. Maps identity
+    /// pubkey associated with the vote account to the validator info.
+    /// Since loading accounts takes a lot of time, we initialize it in a
+    /// separate thread, hence the RwLock - to set the data later from a
+    /// different thread.
+    identity_info_map: RwLock<Option<IdentityInfoMap>>,
+}
+
+impl PrometheusMetrics {
+    pub fn new(
+        bank_forks: Arc<RwLock<BankForks>>,
+        block_commitment_cache: Arc<RwLock<BlockCommitmentCache>>,
+        cluster_info: Arc<ClusterInfo>,
+        vote_accounts: Arc<HashSet<Pubkey>>,
+        snapshot_config: Option<SnapshotConfig>,
+    ) -> Arc<Self> {
+        let prom_metrics = Self {
+            bank_forks: bank_forks.clone(),
+            block_commitment_cache,
+            cluster_info,
+            vote_accounts: vote_accounts.clone(),
+            identity_info_map: RwLock::new(None),
+            snapshot_config,
+        };
+        let prom_metrics = Arc::new(prom_metrics);
+
+        let prom_metrics_clone = prom_metrics.clone();
+        thread::spawn(move || {
+            // TODO: This can panic, we should handle it better
+            let identity_info_map = map_vote_identity_to_info(&bank_forks, &vote_accounts);
+            prom_metrics_clone
+                .identity_info_map
+                .write()
+                .unwrap()
+                .replace(identity_info_map);
+        });
```

Comment:
I would like to understand better why this fixes the problem. Which part of `map_vote_identity_to_info` is taking a long time?

Some candidates:

https://github.com/ChorusOne/solana/blob/f446cf543c3f74d30562f1103e5205975af4f5bd/prometheus/src/identity_info.rs#L30

This is taking the read lock, which may take a long time if the write lock is held while the bank is being initialized. If that is the issue, then:

 * We might have a race, can we be sure that the write lock is already being held when we get here so we get a bank by the time we get the lock? If we get the read lock first, then mapping the vote identities would fail, and also this would delay startup for the thing trying to take the write lock.
 * A nicer solution would be to find the place where the bank is being initialized, and after that code call the thing to initialize the vote identity mapping. Then we have no race and also no need for an additional thread that just sits blocked on the lock.

Though as a quick solution I think the thread is fine.

The other place that could be expensive is https://github.com/ChorusOne/solana/blob/f446cf543c3f74d30562f1103e5205975af4f5bd/prometheus/src/identity_info.rs#L32-L34

because it gets _all_ accounts owned by the config program, which can grow without bounds and it’s in control of users of the network. Probably by now there are thousands of validators more (some of them gone already, but the config accounts continue to exit) than when we initially wrote this. If this is the expensive thing then doing the mapping in a separate thread is a nice solution.

It could also still be this one:
https://github.com/ChorusOne/solana/blob/f446cf543c3f74d30562f1103e5205975af4f5bd/prometheus/src/identity_info.rs#L37

Do you know already which one? If not, we could add logging at each point and see where it spends its time.

Reply:
> `let bank = bank_forks.read().unwrap().working_bank();`

That was my first assumption, and there are some threads using bank forks. I was searching for potential race in the code, however, locks are held very briefly, and even if there are loops acquiring them, I did not find anything that could spin-lock (there are correct sleeps between iterations, without holding the lock).

> let all_accounts = bank 
     .get_program_accounts(&config_program::id(), &ScanConfig::default()) 
     .expect("failed to get program accounts"); 

I am pretty sure it is this one. There is a correlation with startup times taking longer and longer as time goes on, and we also observed growing accounts indexes on RPCs (to the point that we have stopped storing them in ramfs).

Reply:
I have added some logs so that we can confirm or deny my assumption with next release.